shell/platform/darwin/ios/framework/Source/FlutterMetalLayer.mm

```diff
@@ -0,0 +1,388 @@
+
+  // There must be a CADisplayLink scheduled *on main thread* otherwise
+  // core animation only updates layers 60 times a second.
+  CADisplayLink* _displayLink;
+  NSUInteger _displayLinkPauseCountdown;
+
+  // Used to track whether the content was set during this display link.
+  // When unlocking phone the layer (main thread) display link and raster thread
+  // display link get out of sync for several seconds. Even worse, layer display
+  // link does not seem to reflect actual vsync. Forcing the layer link
+  // to max rate (instead range) temporarily seems to fix the issue.
+  BOOL _didSetContentsDuringThisDisplayLinkPeriod;
+
+  // Whether layer displayLink is forced to max rate.
+  BOOL _displayLinkForcedMaxRate;
+}
+
+- (void)presentTexture:(FlutterTexture*)texture;
+- (void)returnTexture:(FlutterTexture*)texture;
+
+@end
+
+@interface FlutterTexture : NSObject {
+  id<MTLTexture> _texture;
+  IOSurface* _surface;
+  CFTimeInterval _presentedTime;
+}
+
+@property(readonly, nonatomic) id<MTLTexture> texture;
+@property(readonly, nonatomic) IOSurface* surface;
+@property(readwrite, nonatomic) CFTimeInterval presentedTime;
+
+@end
+
+@implementation FlutterTexture
+
+@synthesize texture = _texture;
+@synthesize surface = _surface;
+@synthesize presentedTime = _presentedTime;
+
+- (instancetype)initWithTexture:(id<MTLTexture>)texture surface:(IOSurface*)surface {
+  if (self = [super init]) {
+    _texture = texture;
+    _surface = surface;
+  }
+  return self;
+}
+
+@end
+
+@interface FlutterDrawable : NSObject <CAMetalDrawable> {
+  FlutterTexture* _texture;
+  __weak FlutterMetalLayer* _layer;
+  NSUInteger _drawableId;
+  BOOL _presented;
+}
+
+- (instancetype)initWithTexture:(FlutterTexture*)texture
+                          layer:(FlutterMetalLayer*)layer
+                     drawableId:(NSUInteger)drawableId;
+
+@end
+
+@implementation FlutterDrawable
+
+- (instancetype)initWithTexture:(FlutterTexture*)texture
+                          layer:(FlutterMetalLayer*)layer
+                     drawableId:(NSUInteger)drawableId {
+  if (self = [super init]) {
+    _texture = texture;
+    _layer = layer;
+    _drawableId = drawableId;
+  }
+  return self;
+}
+
+- (id<MTLTexture>)texture {
+  return self->_texture.texture;
+}
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunguarded-availability-new"
+- (CAMetalLayer*)layer {
+  return (id)self->_layer;
+}
+#pragma clang diagnostic pop
+
+- (NSUInteger)drawableID {
+  return self->_drawableId;
+}
+
+- (CFTimeInterval)presentedTime {
+  return 0;
+}
+
+- (void)present {
+  [_layer presentTexture:self->_texture];
+  self->_presented = YES;
+}
+
+- (void)dealloc {
+  if (!_presented) {
+    [_layer returnTexture:self->_texture];
+  }
+}
+
+- (void)addPresentedHandler:(nonnull MTLDrawablePresentedHandler)block {
+}
+
+- (void)presentAtTime:(CFTimeInterval)presentationTime {
+}
+
+- (void)presentAfterMinimumDuration:(CFTimeInterval)duration {
+}
```

Comment:
For methods that aren't implemented, would it be safer to error log if they are used?